### PR TITLE
feat: [EXAM-1377] adds hardware checker templates for microphone and …

### PR DESCRIPTION
**Pull request by Golub-Sergey**
**Description:** adds hardware checker templates for microphone and camera

**Youtrack:** https://youtrack.raccoongang.com/issue/EXAM-1377

**Reviewers:**
- [ ] @nikolay.borovenskiy

**Merge checklist:**
- [ ] Demo status: OK
- [ ] All related documentation is updated
- [ ] Commits are squashed
- [ ] DevOps notes added

**Post merge:**
- [ ] Delete the working branch (if not needed anymore)

### CODE REVIEW

File: edx_proctoring/api.py [hunk 1]
```diff
@@ -78,6 +80,13 @@
 FIX_ATTEMPT_IN_COMPLETED_STATE_DELAY_MINS = 30  # it is necessary to end timed exam properly
 
 
+CHECK_HARDWARE_TEMPLATE = {
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/api.py]: templates_map

File: edx_proctoring/api.py [hunk 2]
```diff
@@ -1687,13 +1696,29 @@ def _get_timed_exam_view(exam, context, exam_id, user_id, course_id):
...
+        )
         if is_exam_passed_due(exam, user=user_id):
             student_view_template = 'timed_exam/expired.html'
         elif (
-                context.get('is_checking_audio_enabled') or
-                context.get('is_checking_microphone_enabled')
+                'hardware_checker' in get_current_request().GET.keys() and
+                any(enabled_hardwares) and
+                CourseBlockHardwareChecker.are_hardwares_enabled(
+                    user_id, exam['content_id'],
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/api.py]: Bad formating.

File: edx_proctoring/api.py [hunk 3]
```diff
@@ -1687,13 +1696,29 @@ def _get_timed_exam_view(exam, context, exam_id, user_id, course_id):
...
+                ('is_checking_camera_enabled', context.get('is_checking_camera_enabled')),
+            )
+        )
         if is_exam_passed_due(exam, user=user_id):
             student_view_template = 'timed_exam/expired.html'
         elif (
-                context.get('is_checking_audio_enabled') or
-                context.get('is_checking_microphone_enabled')
+                'hardware_checker' in get_current_request().GET.keys() and
+                any(enabled_hardwares) and
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/api.py]: Do not understand `any` here.

File: edx_proctoring/api.py [hunk 4]
```diff
@@ -1687,13 +1696,29 @@ def _get_timed_exam_view(exam, context, exam_id, user_id, course_id):
     attempt_status = attempt['status'] if attempt else None
     has_due_date = exam['due_date'] is not None
     if not attempt_status:
+        enabled_hardwares = tuple(
+            hardware_name
+            for hardware_name, value in (
+                ('is_checking_audio_enabled', context.get('is_checking_audio_enabled')),
+                ('is_checking_microphone_enabled', context.get('is_checking_microphone_enabled')),
+                ('is_checking_camera_enabled', context.get('is_checking_camera_enabled')),
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/api.py]: U use a lot of plain stings `is_checking_audio_enabled`, `is_checking_audio_enabled`, `is_checking_audio_enabled` in your feature. It should be stored somewhere in one place. In a class or somewhere else. But you need to try to remove copy-paste and make the code reliable.

File: edx_proctoring/api.py [hunk 5]
```diff
@@ -1687,13 +1696,29 @@ def _get_timed_exam_view(exam, context, exam_id, user_id, course_id):
     attempt_status = attempt['status'] if attempt else None
     has_due_date = exam['due_date'] is not None
     if not attempt_status:
+        enabled_hardwares = tuple(
+            hardware_name
+            for hardware_name, value in (
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/api.py]: Do you need value here?

File: edx_proctoring/api.py [hunk 6]
```diff
@@ -1687,13 +1696,29 @@ def _get_timed_exam_view(exam, context, exam_id, user_id, course_id):
     attempt_status = attempt['status'] if attempt else None
     has_due_date = exam['due_date'] is not None
     if not attempt_status:
+        enabled_hardwares = tuple(
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/api.py]: Enabled? Maybe available?

File: edx_proctoring/models.py [hunk 7]
```diff
@@ -820,3 +820,46 @@ class Meta:
...
+    'microphone': 'is_checking_microphone_enabled',
+    'headphones': 'is_checking_audio_enabled',
+    'camera': 'is_checking_camera_enabled',
+}
+
+
+class CourseBlockHardwareChecker(models.Model):
+    """
+    Store hardware checking progress.
+    """
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/models.py:835]: Add one blank line after

File: edx_proctoring/models.py [hunk 8]
```diff
@@ -820,3 +820,46 @@ class Meta:
...
+    user_id = models.PositiveIntegerField()
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/models.py:836]: Not foreign key? Are you sure?

File: edx_proctoring/models.py [hunk 9]
```diff
@@ -820,3 +820,46 @@ class Meta:
...
+
+
+HARDWARE_WAS_CHECKED = {
+    'microphone': 'is_checking_microphone_enabled',
+    'headphones': 'is_checking_audio_enabled',
+    'camera': 'is_checking_camera_enabled',
+}
+
+
+class CourseBlockHardwareChecker(models.Model):
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/models.py:832]: What does it mean CourseBlock?

File: edx_proctoring/models.py [hunk 10]
```diff
@@ -820,3 +820,46 @@ class Meta:
...
+    """
+    Store hardware checking progress.
+    """
+    user_id = models.PositiveIntegerField()
+    content_id = models.CharField(max_length=100)
+    is_checking_audio_enabled = models.BooleanField(default=False)
+    is_checking_microphone_enabled = models.BooleanField(default=False)
+    is_checking_camera_enabled = models.BooleanField(default=False)
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/models.py:840]: Maybe it is better to use `passed` instead of `enabled`?

File: edx_proctoring/models.py [hunk 11]
```diff
@@ -820,3 +820,46 @@ class Meta:
...
+
+    @classmethod
+    def are_hardwares_enabled(
+        cls,
+        user_id,
+        content_id,
+        enabled_hardwares,
+        hardware_checked=''
+    ):
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/models.py:849]: Please typing do not forget.

File: edx_proctoring/models.py [hunk 12]
```diff
@@ -820,3 +820,46 @@ class Meta:
...
+        hardware, __ = cls.objects.get_or_create(
+            defaults={
+                'user_id': user_id,
+                'content_id': content_id,
+            }
+        )
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/models.py]: What? The model method class, which must return yes or no, creates an instance of this model. This is a huge side effect. You can't do that.

File: edx_proctoring/models.py [hunk 13]
```diff
@@ -820,3 +820,46 @@ class Meta:
...
+
+        if hardware_checked in HARDWARE_WAS_CHECKED.keys():
+            setattr(hardware, HARDWARE_WAS_CHECKED[hardware_checked], True)
+            hardware.save()
+
+        for hardware_name in enabled_hardwares:
+            if not getattr(hardware, hardware_name):
+                return True
+
+        return False
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/models.py]: `return any(not getattr(hardware, hardware_name) for hardware_name in enabled_hardwares))`
?

File: edx_proctoring/models.py [hunk 14]
```diff
@@ -820,3 +820,46 @@ class Meta:
...
+
+    @classmethod
+    def are_hardwares_enabled(
+        cls,
+        user_id,
+        content_id,
+        enabled_hardwares,
+        hardware_checked=''
+    ):
+        hardware, __ = cls.objects.get_or_create(
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/models.py]: Just `hardware` is a bad name.

File: edx_proctoring/templates/timed_exam/check_camera_hardware.html [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+{% include 'timed_exam/entrance.html' %}
+
+<script type="text/javascript">
+
+  $('.camera-problem-solving-tips').hide();
+  $('.entrance').hide();
+  $('.footer-sequence').hide();
+
+  $('.btn-success-checking-hardware').click(function () {
+    var msg = "Ви впевнені, що чули звукозапис?";
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/templates/timed_exam/check_camera_hardware.html]: Did you have a problem with translation?

File: edx_proctoring/templatetags/template_helpers.py [hunk 16]
```diff
@@ -0,0 +1,16 @@
+"""
+Provide template helpers method for proctoring templates.
+"""
+from django import template
+from exam_dashboard.utils import extend_url_with_query_parameters as modify_url_with_query_parameters
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/templatetags/template_helpers.py]: The proctoring currently uses the `exam_dashboard` package as a dependency. This is bad. I propose to move the `extend_url_with_query_parameters` function to the common utilities and store it there and import it from there already.

File: edx_proctoring/templatetags/template_helpers.py [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+
+register = template.Library()
+
+
+@register.simple_tag
+def extend_url_with_query_parameters(url, hardware_checker, hardware_checked):
```
Comment (NikolayBorovenskiy) [inline edx_proctoring/templatetags/template_helpers.py]: Haven't you done a tag like this before?

Comment (NikolayBorovenskiy) [inline edx_proctoring/templatetags/template_helpers.py]: Please add typing. Always add typing 🙂

**Review by NikolayBorovenskiy — CHANGES_REQUESTED** [review]

File: edx_proctoring/templatetags/template_helpers.py (hunk 17 above)
Comment (Golub-Sergey) [reply edx_proctoring/templatetags/template_helpers.py]: No, I haven't done it

**Review by NikolayBorovenskiy — APPROVED** [review]

**Pull request merged**